File: android_mobile/src/main/java/com/alexstyl/specialdates/contact/ContactsProvider.java
package com.alexstyl.specialdates.contact;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class ContactsProvider {

    private final Map<Integer, ContactsProviderSource> sources;

    ContactsProvider(Map<Integer, ContactsProviderSource> sources) {
        this.sources = sources;
    }

    public Contacts getContacts(List<Long> contactIds, @ContactSource int source) {
        if (sources.containsKey(source)) {
            return sources.get(source).queryContacts(contactIds);
        }
        throw new IllegalArgumentException("Unknown source type: " + source);
    }

    public Contact getContact(long contactID, @ContactSource int source) throws ContactNotFoundException {
        if (sources.containsKey(source)) {
            return sources.get(source).getOrCreateContact(contactID);
        }
        throw new IllegalArgumentException("Unknown source type: " + source);

    }

    public List<Contact> getAllContacts() {
        List<Contact> contacts = new ArrayList<>();
        for (ContactsProviderSource providerSource : sources.values()) {
            contacts.addAll(providerSource.getAllContacts().getContacts());
        }
        return Collections.unmodifiableList(contacts);
    }

}

File: memento/src/main/java/com/alexstyl/specialdates/contact/ContactsProvider.kt
package com.alexstyl.specialdates.contact

class ContactsProvider(private val sources: Map<Int, ContactsProviderSource>) {

    val allContacts: List<Contact>
        get() = sources.values.fold(emptyList()) { list, source ->
            list + source.allContacts
        }

    fun getContacts(contactIds: List<Long>, @ContactSource source: Int): List<Contact> {
        return sources[source]!!.queryContacts(contactIds)
    }

    @Throws(ContactNotFoundException::class)
    fun getContact(contactID: Long, @ContactSource source: Int): Contact {
        return sources[source]!!.getOrCreateContact(contactID)
    }

}
